src/debug.c

```diff
@@ -2477,45 +2473,75 @@ void debugDelay(int usec) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
 /** If it doesn't block and doesn't ignore, return 1 (the thread will handle the signal) 
  * If thread tid blocks or ignores sig_num returns 0 (thread is not ready to catch the signal).
- * also returns 0 if it failed to open the thread's status file. **/
+ * also returns 0 if something is wrong and prints a warnning message to the log file **/
 static int is_thread_ready_to_signal(pid_t pid, pid_t tid, int sig_num) {
     /* open the threads status file */
     char buff[MAX_BUFF_LENGTH];
     snprintf(buff, MAX_BUFF_LENGTH, "/proc/%d/task/%d/status", pid, tid);
     FILE *thread_status_file = fopen(buff, "r");
     if (thread_status_file == NULL) {
+        serverLog(LL_WARNING,
+        "tid:%d: failed to open /proc/%d/task/%d/status file", tid, pid, tid);
         return 0;
     }
 
     int ret = 1;
-
-    size_t field_name_len = strlen("SigBlk:");
-    while (fgets(buff, MAX_BUFF_LENGTH, thread_status_file)) {
-        /* iterate the file until we reach SigBlk field line*/
-        if (!strncmp(buff, "SigBlk:", field_name_len)) {
-            /* check if the signal is blocked by this thread */
-            unsigned long sig_block_mask = strtoul(buff + field_name_len, NULL, 16);
-            if(sig_block_mask & sig_num) {
+    size_t field_name_len = strlen("SigBlk:"); /* SigIgn has the same length */
+    size_t list_len = 2;
+    status_file_field_attr fields_list[] = {{.name = "SigBlk:", .name_len = field_name_len},
+                                            {.name = "SigIgn:", .name_len = field_name_len}};
+    char *line = NULL;
+    while ((line = fgets(buff, MAX_BUFF_LENGTH, thread_status_file))) {
+        /* iterate the file until we reach SigBlk or SigIgn field line*/
+        if (line_in_fields_list(buff, fields_list, list_len)) {
+            /* check if the signal exist in the mask */
+            unsigned long sig_mask = strtoul(buff + field_name_len, NULL, 16);
+            if(sig_mask & sig_num) { /* if the signal is blocked/ignored return 0*/
                 ret = 0;
-            } else {
-                /* get SigIgn field line and check if this thread ignores sig_num*/
-                if (fgets(buff, MAX_BUFF_LENGTH, thread_status_file)) {/* Avoid warning. */};
-                unsigned long sig_ign_mask = strtoul(buff + field_name_len, NULL, 16);
-                if(sig_ign_mask & sig_num) {
-                    ret = 0;
-                }
+                break;
             }
 
-            break;
+            /* we found the second field and it doesn't include sig_num as well. stop seraching.*/
+            if (list_len == 1) break;
+            
+            /* if we found SigBlk, serach for SigIgn and vice versa. */
+            const char *field = !strncmp(buff, "SigBlk:", field_name_len) ? "SigIgn:" : "SigBlk:";
+            fields_list[0].name = field;
+            list_len = 1;
```

Comment:
i find this logic of this function a bit odd (even though it's probably efficient).
maybe it'll be nicer to extract each of these two fields into a variable, and put the logic that looks at the values of these variables after.

maybe without a dedicated struct and a dedicated function.
like we do argument parsing of a command, e.g. look at `geoaddCommand`.
we can break out as soon as we have values for both, or even as soon as we have value for either.

WDYT?

Reply:
My intention was to to discard one `strncmp` after finding one of the fields (unlike `geoaddCommand`)
and to make it easier to generalize `is_thread_ready_to_signal` in the future if we decide we want to extract more fields from the file. 
i can do
```
    size_t requested_fields = 2;
    while ((line = fgets(buff, MAX_BUFF_LENGTH, thread_status_file)) && requested_fields) {
        /* iterate the file until we reach SigBlk or SigIgn field line*/
        if (!strncmp(buff, "SigBlk:", field_name_len) ||  !strncmp(buff, "SigIgn:", field_name_len)) {
            /* check if the signal exist in the mask */
            unsigned long sig_mask = strtoul(buff + field_name_len, NULL, 16);
            if(sig_mask & sig_num) { /* if the signal is blocked/ignored return 0*/
                ret = 0;
                break;
            }

            --requested_fields;
        }
    }
```
This is probably cleaner and considering the fact that SigBlk and SigIgn are probably sequential, it won't affect the performance. But less general

Reply:
i'd rather that version, doesn't attempt to create some common infrastructure for something that's very specific